internal/uvm/virtual_device.go

```diff
@@ -15,54 +15,106 @@ import (
+// AssignDeviceToUVM assigns a vpci device to the uvm
+// if the device already exists, the stored VPCIDevice's ref count is increased
+// and the VPCIDevice is returned.
+// Otherwise, a new request is made to assign the target device indicated by the deviceID
+// onto the UVM. A new VPCIDevice entry is made on the UVM and the VPCIDevice is returned
+// to the caller
+func (uvm *UtilityVM) AssignDeviceToUVM(ctx context.Context, deviceID string) (*VPCIDevice, error) {
 	guid, err := guid.NewV4()
 	if err != nil {
 		return nil, err
 	}
-	id := guid.String()
+	vmBusGUID := guid.String()
 
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	if err := uvm.modify(ctx, &hcsschema.ModifySettingRequest{
-		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, id),
+
+	existingVPCIDevice := uvm.vpciDevices[deviceID]
+	if existingVPCIDevice != nil {
+		existingVPCIDevice.refCount++
+		return existingVPCIDevice, nil
+	}
+
+	targetDevice := hcsschema.VirtualPciDevice{
+		Functions: []hcsschema.VirtualPciFunction{
+			{
+				DeviceInstancePath: deviceID,
+			},
+		},
+	}
+
+	request := &hcsschema.ModifySettingRequest{
+		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, vmBusGUID),
 		RequestType:  requesttype.Add,
-		Settings:     device,
-		GuestRequest: guestrequest.GuestRequest{
+		Settings:     targetDevice}
+
+	// we do not currently support assigned devices on windows,
+	// so no guest side work necessary for windows
+	if uvm.operatingSystem != "windows" {
+		// for LCOW, we need to make sure that specific paths relating to the
+		// device exist so they are ready to be used by later
+		// work in openGCS
+		request.GuestRequest = guestrequest.GuestRequest{
 			ResourceType: guestrequest.ResourceTypeVPCIDevice,
 			RequestType:  requesttype.Add,
 			Settings: guestrequest.LCOWMappedVPCIDevice{
-				VMBusGUID: id,
+				VMBusGUID: vmBusGUID,
 			},
-		},
-	}); err != nil {
+		}
+	}
+
+	if err := uvm.modify(ctx, request); err != nil {
 		return nil, err
 	}
-	return &VPCIDevice{
-		vm: uvm,
-		ID: id,
-	}, nil
+	result := &VPCIDevice{
+		vm:               uvm,
+		VMBusGUID:        vmBusGUID,
+		deviceInstanceID: deviceID,
+		refCount:         1,
+	}
+	uvm.vpciDevices[deviceID] = result
+	return result, nil
 }
 
-// RemoveDevice removes a vpci device from the uvm
-func (uvm *UtilityVM) RemoveDevice(ctx context.Context, id string) error {
+// RemoveDeviceFromUVM removes a vpci device from a uvm when there are
+// no more references to a given VPCIDevice. Otherwise, decrements
+// the reference count of the VPCIDevice and returns nil.
+func (uvm *UtilityVM) RemoveDeviceFromUVM(ctx context.Context, vpci *VPCIDevice) error {
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	return uvm.modify(ctx, &hcsschema.ModifySettingRequest{
-		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, id),
-		RequestType:  requesttype.Remove,
-	})
+
+	device := uvm.vpciDevices[vpci.deviceInstanceID]
+	if device == nil {
+		return nil
```

Comment:
Should we return an error instead in this case? What do our other resource types do if you try to release something that was never added?

Reply:
I'm fine logging an error, but if we return an error here, then all resource cleanup will stop and we may end up leaking resources.

Reply:
Not the case anymore :). ReleaseResources continues to release until after the loop and then throws an error if any of the clean ups failed to avoid this now.

Reply:
If you haven't rebased to get the GMSA stuff yet it's in there

Reply:
Done.